devito/tools/utils.py

```diff
@@ -165,7 +165,14 @@ def filter_ordered(elements, key=None):
             return unordered[np.argsort(inds)].tolist()
         except:
             return sorted(list(set(elements)), key=elements.index)
-    return [e for e in elements if not (key(e) in seen or seen.add(key(e)))]
+    else:
+        ret = []
+        for e in elements:
+            k = key(e)
+            if k not in seen:
+                ret.append(e)
+                seen.add(k)
+        return ret
```

Comment:
A faster filtering method (e.g. using `set`) is not in some way applicable here?

Reply:
elaborate? we need to preserve ordering btw

Reply:
If ordering needs to be preserved then it won't make much difference, but a `sorted(set(mylist), key=mylist.index)` structure type would be a little quicker?

Reply:
so the filtering is dictated by the `key`, but in your solution you're filtering based on the elements' `__eq__` (implicitly through `set`), so the outcome would be different right?